File: CodeWars/Fundamentals/ColorChoice/ColorChoice.c
#include <stdio.h>
#include <assert.h>

long long binCoef (long long n, int x) {
    long long res = 1;
    if (x > n - x)
        x = n - x;

    for (int i = 0 ; i < x ; ++i) {
        res *= (n - i);
        res /= (i + 1);
    }
    return res;
}

long long checkChoose (long long m, int n) {
    for (int i = 0 ; i < n ; ++i) {
        if (binCoef(n, i) == m)
            return i;
    }
    return -1;
}

void tests () {
    assert (checkChoose(6, 4) == 2);
    assert(checkChoose(4, 4) == 1);
    assert(checkChoose(4, 2) == -1);
    assert(checkChoose(35, 7) == 3);
    assert(checkChoose(155117520, 30) == 15);

}

int main () {
    tests ();
    return 0;
}

File: CodeWars/Fundamentals/ColorChoice/ColorChoice.rs
fn bin_coef(n: u64, x: u64) -> u64 {
    let mut x = x;
    let mut res = 1;

    if x > (n-x) {
        x = n-x;
    }

    for i in 0..x {
        res *= n-i;
        res /= i+1;
    }

    res
}

fn check_choose(m: u64, n: u64) -> i64 {
    for i in 0..n {
        if bin_coef(n, i) == m {
            return i as i64;
        }
    }

    -1
}

fn dotest(m: u64, n: u64, exp: i64) -> () {
    assert_eq!(check_choose(m, n), exp)
}

fn basics_check_choose() {
    dotest(6, 4, 2);
    dotest(4, 4, 1);
    dotest(35, 7, 3);
    dotest(36, 7, -1);
    dotest(184756, 20, 10);
}

fn main() {
    basics_check_choose();
}
